x-wing/Cargo.toml

```diff
@@ -0,0 +1,35 @@
+[package]
+name = "x-wing"
+description = """
+Pure Rust implementation of the X-Wing Key-Encapsulation Mechanism
+"""
+version = "0.1.0"
+edition = "2021"
+rust-version = "1.81"
+license = "Apache-2.0 OR MIT"
+readme = "README.md"
+homepage = "https://github.com/RustCrypto/KEMs/tree/master/ml-kem"
+repository = "https://github.com/RustCrypto/KEMs"
+categories = ["cryptography", "no-std"]
+keywords = ["crypto", "x-wing", "xwing", "kem", "post-quantum"]
+
+[dependencies]
+rand_core = { version = "0.6.4", default-features = false }
+x25519-dalek = { version = "2.0", features = [
```

Comment:
You might consider using `curve25519-dalek` directly in order to have one fewer dependency.

`x25519-dalek` is more of a high-level API aimed at ECDH use cases.

You can use the `Scalar` and `MontgomeryPoint` types in lieu of `EphemeralSecret` and `PublicKey` respectively.

Reply:
Think this also means I would need to implement the `x25519` function as I can't find it in `curve25519-dalek`, although that function is surprisingly simple. I'll have a look if it maters mutch in the dependency graph.

Reply:
It's just scalar multiplication, which you can do with the `Mul` trait. Here's an example: https://github.com/RustCrypto/nacl-compat/blob/master/crypto_box/src/lib.rs#L263

Reply:
Also it's fine to do this as a followup, or I can do it if you'd like. It would be good to get an initial working version of this out.

Reply:
I think `x25519-dalek` might be linking `std` though, which is breaking the tests. Can you at least add `default-features = false`?

Reply:
> It's just scalar multiplication, which you can do with the `Mul` trait. Here's an example: https://github.com/RustCrypto/nacl-compat/blob/a7f32e4/crypto_box/src/lib.rs#L263

The implementation in x25519-dalek is indeed just a multiplication. Though its clamped, where in the implementation you linked its just a multiplication.
```
pub fn x25519(k: [u8; 32], u: [u8; 32]) -> [u8; 32] {
    MontgomeryPoint(u).mul_clamped(k).to_bytes()
}
```
Removing the dependency would also require implementing the `x25519_dalek::PublicKey::from(&sk_x);` function to construct the public key.

Removing the dependency will just remove the direct dependency, no transitive dependencies will be dropped. So I wonder if its worth the effort. Did minimize the features.
```
x-wing v0.1.0
├── ml-kem v0.2.1
│   ├── hybrid-array v0.2.0-rc.9
│   │   └── typenum v1.17.0
│   ├── kem v0.3.0-pre.0
│   │   ├── rand_core v0.6.4
│   │   │   └── getrandom v0.2.12
│   │   │       ├── cfg-if v1.0.0
│   │   │       └── libc v0.2.153
│   │   └── zeroize v1.8.1
│   ├── rand_core v0.6.4 (*)
│   └── sha3 v0.10.8
│       ├── digest v0.10.7
│       │   ├── block-buffer v0.10.4
│       │   │   └── generic-array v0.14.7
│       │   │       └── typenum v1.17.0
│       │   │       [build-dependencies]
│       │   │       └── version_check v0.9.4
│       │   └── crypto-common v0.1.6
│       │       ├── generic-array v0.14.7 (*)
│       │       └── typenum v1.17.0
│       └── keccak v0.1.5
├── rand_core v0.6.4 (*)
├── sha3 v0.10.8 (*)
└── x25519-dalek v2.0.1
    ├── curve25519-dalek v4.1.2
    │   ├── cfg-if v1.0.0
    │   ├── cpufeatures v0.2.12
    │   ├── curve25519-dalek-derive v0.1.1 (proc-macro)
    │   │   ├── proc-macro2 v1.0.78
    │   │   │   └── unicode-ident v1.0.12
    │   │   ├── quote v1.0.35
    │   │   │   └── proc-macro2 v1.0.78 (*)
    │   │   └── syn v2.0.52
    │   │       ├── proc-macro2 v1.0.78 (*)
    │   │       ├── quote v1.0.35 (*)
    │   │       └── unicode-ident v1.0.12
    │   └── subtle v2.5.0
    │   [build-dependencies]
    │   ├── platforms v3.4.0
    │   └── rustc_version v0.4.0
    │       └── semver v1.0.22
    └── rand_core v0.6.4 (*)
    ```

Reply:
I can take care of it in a followup.

You don't need `x25519_dalek::PublicKey`... it's just a wrapper for `MontgomeryPoint`.

Clamping actually applies to the private scalar. One way or another it's a scalar multiplication, however `mul_clamped` can be handy if you can't ensure the private scalar is clamped to begin with.

Reply:
All the ECC calculations are new for me so I'm a bit carefull with changes like this. And figured I would first look at the kem traits / array types.